### DISCUSSION TITLE
How to deal with "java.util.zip.ZipException: Not in GZIP format" exception while processing request

### DISCUSSION THREAD

**Discussion by DorisQQ2019** [Kong/unirest-java#427] [Q&A] [answered]
Hello! I receive "com.mashape.unirest.http.exceptions.UnirestException: java.lang.RuntimeException: java.lang.RuntimeException: java.util.zip.ZipException: Not in GZIP format " exception when processing request (request.asString() and request.asBinary() ), however I did not see any exception while using curl. It seems, there is something wrong within response header which causes the exception. According to some hints in website (https://stackoverflow.com/questions/13018732/java-util-zip-zipexception-not-in-gzip-format), we could reset header property to solve this issue. My question is, how to reset this using unirest?

**Comment by ryber** [comment]
So a few things, 
1. You are on a very very old version of Unirest and the handling of GZIP is a little different now. So I would try upgrading to the new kong version and see if that helps.
2. I don't think there is a good way to re-write the RESPONSE header telling you the encoding. The newest version of Unirest has a method (```<T> HttpResponse<T> asObject(Function<RawResponse, T> function)``` which I think would let you read the bytes raw and do the encoding yourself.

**Comment by DorisQQ2019** [comment]
thanks, I did try as above mentioned, however, here is a further question, no mater what stream I use to read data, the "not in gzip format" exception occurrs. Even I invoke rawResponse.getContentAsBytes() , the exception occurrs the same. It seems, I did not get the real raw data. I debug the raw response, it shows me gzipinputstream as below:
![image](https://user-images.githubusercontent.com/57432120/153714679-7d59dd5f-1dd2-40b4-ac3d-fa9314830cd3.png)

So the question now is, how can I get the raw response and decode the content?
BTW, I am using 3.9.00 version to be compatible with my project code

**Comment by ryber** [comment, marked as answer]
You can get ahold of the raw Apache InputStream when using the ```thenConsume``` method. This is before Unirest has read it, so it's the raw socket.

```java
Unirest.get(path).thenConsume(raw -> {
            InputStream i = raw.getContent();
        });
```

once you have read the InputStream it can't be read again, which is why this method is a consumer. 
Your real problem is that the API is reporting the content is gzipped but it's not. Have you reported this to the API owner?

**Reply by DorisQQ2019** [reply]
Tahnks for your replay. I will try it. 
The API owner knew this problem, for some reason, he did not have plan to fix it.

**Reply by DorisQQ2019** [reply]
![image](https://user-images.githubusercontent.com/57432120/153718557-dcdc7130-b094-416c-a63c-eff8f909c20d.png)
Unfortunately , the original data stream has been categorized as gzipinstream in comsumer...

**Reply by ryber** [reply]
Ah, well thats coming from Apache, before Unirest even gets ahold of it.

```java
Unirest.config().httpClient(ApacheClient.builder().withRequestConfig((config, httpRequest) -> {
            return RequestConfig.custom().setContentCompressionEnabled(false).build();
        }));
```

Now once that is done I'm not exactly sure of the behavior, When apache does the decoding it actually removes the gzip encoding from the response back to unirest. If unirest sees the gzip header it will also try and unzip it. This is because Apache only unzips on the synchronous client and not the async one for 'reasons". So, with the encoding in Apache turned off, you may still need to use the thenConsume, but you could try other methods they might work.

**Reply by ryber** [reply]
oh and if you do that config, you're going to want to do that ONCE, don't do it before each request because building a client is a big deal and starts watcher threads and stuff.

**Reply by DorisQQ2019** [reply]
Unirest.config().httpClient(ApacheClient.builder(HttpClients.createDefault()).withRequestConfig((config, httpRequest) -> {
            return RequestConfig.custom().setContentCompressionEnabled(false).build();
        }));

Thank you, this piece of code works, I did not see the exception any more.
